cupy/_core/dlpack.pyx

```diff
@@ -303,3 +309,58 @@ cpdef ndarray fromDlpack(object dltensor) except +:
     # Make sure this capsule will never be used again.
     cpython.PyCapsule_SetName(mem.dltensor, 'used_dltensor')
     return ndarray(shape_vec, cp_dtype, mem_ptr, strides=strides_vec)
+
+
+# TODO(leofang): this function is exposed to the cupy namespace, so it returns
+# a cupy.ndarray which is not compliant with the Python array API. When we have
+# a compliant object living in, say, cupy.array_api, we will expose another
+# function cupy.array_api.from_dlpack().
+cpdef from_dlpack(array):
+    """Zero-copy conversion between array objects compliant with the DLPack
+    data exchange protocol.
+
+    Args:
+        array (object): an array object that implements two methods:
+            ``__dlpack__()`` and ``__dlpack_device__()``.
+
+    Returns:
+        cupy.ndarray: a CuPy array that can be safely accessed on CuPy's
+            current stream.
+
+    .. note::
+        This function is different from CuPy's legacy :func:`~cupy.fromDlpack`
+        function in that the latter takes a :class:`PyCapsule` object that
+        contains the DLPack tensor as input, while the former takes any object
+        implementing the DLPack data exchange protocol.
+
+    .. seealso::
+        `Data interchange mechanisms`_
+
+    .. _Data interchange mechanisms:
+        https://data-apis.org/array-api/latest/design_topics/data_interchange.html
+    """
+    try:
+        dev_type, dev_id = array.__dlpack_device__()
```

Comment:
Could you use `hasattr` instead? In general, it's better to avoid relying on exceptions in normal cases. This also avoids catching `AtttributeError` raised from inside `__dlpack_device__()`.

Reply:
I think `__dlpack_device__` shouldn't raise, but ok will do.